pkg/cloudprovider/openstack.go

```diff
@@ -0,0 +1,728 @@
+	// cloudprivateipconfig objects are created simultaneously for the exact same server.
+	// IMO the only solution is extending gophercloud with the revision number mechanism and then
+	// relying on that.
+	_, err := ports.Update(o.neutronClient, p.ID, opts).Extract()
+	return err
+}
+
+// getNeutronSubnetsForNetwork returns all subnets that belong to the given network with
+// ID <networkID>.
+func (o *OpenStack) getNeutronSubnetsForNetwork(networkID string) ([]neutronsubnets.Subnet, error) {
+	var subnets []neutronsubnets.Subnet
+
+	if _, err := uuid.Parse(networkID); err != nil {
+		return nil, fmt.Errorf("networkID '%s' is not a valid UUID", networkID)
+	}
+
+	opts := neutronsubnets.ListOpts{NetworkID: networkID}
+	pager := neutronsubnets.List(o.neutronClient, opts)
+	err := pager.EachPage(func(page pagination.Page) (bool, error) {
+		subnetList, err := neutronsubnets.ExtractSubnets(page)
+		if err != nil {
+			return false, err
+		}
+		for _, s := range subnetList {
+			subnets = append(subnets, s)
+		}
+		return true, nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	return subnets, nil
+}
+
+// getNovaServer gets the nova server with ID == <serverID>.
+func (o *OpenStack) getNovaServer(serverID string) (*novaservers.Server, error) {
+	if _, err := uuid.Parse(serverID); err != nil {
+		return nil, fmt.Errorf("serverID '%s' is not a valid UUID", serverID)
+	}
+
+	server, err := novaservers.Get(o.novaClient, serverID).Extract()
+	if err != nil {
+		return nil, err
+	}
+	return server, nil
+}
+
+// listNovaServerPorts lists all ports that are attached to the provided nova server
+// with ID == <serverID>.
+func (o *OpenStack) listNovaServerPorts(serverID string) ([]neutronports.Port, error) {
+	var err error
+	var serverPorts []neutronports.Port
+
+	if _, err := uuid.Parse(serverID); err != nil {
+		return nil, fmt.Errorf("serverID '%s' is not a valid UUID", serverID)
+	}
+
+	portListOpts := neutronports.ListOpts{
+		DeviceID: serverID,
+	}
+
+	pager := neutronports.List(o.neutronClient, portListOpts)
+	err = pager.EachPage(func(page pagination.Page) (bool, error) {
+		portList, err := neutronports.ExtractPorts(page)
+		if err != nil {
+			return false, err
+		}
+		for _, p := range portList {
+			serverPorts = append(serverPorts, p)
+		}
+		return true, nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	return serverPorts, nil
+}
+
+// isIPAddressAllowedOnNeutronPort returns true if the given IP address can be found inside the
+// list of allowed_address_pairs for this port.
+func isIPAddressAllowedOnNeutronPort(p neutronports.Port, ip net.IP) bool {
+	for _, aap := range p.AllowedAddressPairs {
+		if ip.Equal(net.ParseIP(aap.IPAddress)) {
+			return true
+		}
+
+	}
+	return false
+}
+
+// hasNodeName will report true if the given description can be parsed correctly and
+// if given nodeName can be found inside the description.
+func hasNodeName(description string, nodeName string) bool {
```

Comment:
NIT: how about making this return (bool, error) in case we care to know a difference between unmarshal failures from values mismatch?

Reply:
done